### PR TITLE
Refactor the GraphImpl portion of Universe into Pathfinder.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
Separate the functionality encompassed by the graph of starlanes, and pathfinding into its own class.

The benefits will include the following.

Better code partitioning and a smaller Universe class.

Faster partial compile times since Pathfinder is a pimpl implementation.

The ability to optimize the pathfinding functions, in the immediate future the neighborhood region finding functions, which is currently prevented because the distance_matrix_storage and the implementation
of the distance cache are on the different sides of the public private
divide of Universe.h and Universe.cpp.

### CODE REVIEW

File: universe/Pathfinder.h [hunk 1]
```diff
@@ -0,0 +1,148 @@
+#ifndef _Pathfinder_h_
+#define _Pathfinder_h_
+
+#include "Enums.h"
```
Comment (geoffthemedio) [inline universe/Pathfinder.h]: are Enums.h and ObjectMap.h needed in the header?

File: universe/Pathfinder.h [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+#include <map>
+#include <set>
+
+#ifdef FREEORION_WIN32
+// because the linker gets confused about Win32 API functions...
+#  undef GetObject
+#  undef GetObjectA
+#endif
+
+struct UniverseObjectVisitor;
```
Comment (geoffthemedio) [inline universe/Pathfinder.h]: are all these forward declares needed?

File: universe/Universe.cpp [hunk 3]
```diff
@@ -83,345 +84,11 @@ namespace boost {
...
-        void non_tree_edge(const Edge& e, const Graph& g) {}
-        void gray_target(const Edge& e, const Graph& g) {}
-        void black_target(const Edge& e, const Graph& g) {}
-        void finish_vertex(const Vertex& e, const Graph& g) {}
-    };
-
-    ////////////////////////////////////////////////////////////////
-    // templated implementations of Universe graph search methods //
-    ////////////////////////////////////////////////////////////////
-    struct vertex_system_id_t {typedef boost::vertex_property_tag kind;}; ///< a system graph property map type
```
Comment (geoffthemedio) [inline universe/Universe.cpp:174]: presumably a header or three can be removed from Universe.cpp if uses of this sort of type are removed from the file...

File: util/Order.cpp [hunk 4]
```diff
@@ -277,7 +278,7 @@ FleetMoveOrder::FleetMoveOrder(int empire, int fleet_id, int start_system_id, in
         return;
     }
 
-    std::pair<std::list<int>, double> short_path = GetUniverse().ShortestPath(m_start_system, m_dest_system, empire);
+    std::pair<std::list<int>, double> short_path = GetUniverse().GetPathfinder()->ShortestPath(m_start_system, m_dest_system, empire);
```
Comment (geoffthemedio) [inline util/Order.cpp]: rather than two layers of getters everywhere, maybe have a free GetPathfinder() function available?

**Comment by adrianbroher** [comment]
Why there are refactoring commit in new code?  Please fixup those into the corresponding code.

File: universe/Pathfinder.h (hunk 1 above)
Comment (LGM-Doyle) [reply universe/Pathfinder.h]: Fixed.

File: universe/Pathfinder.h (hunk 2 above)
Comment (LGM-Doyle) [reply universe/Pathfinder.h]: Fixed.

File: util/Order.cpp (hunk 4 above)
Comment (LGM-Doyle) [reply util/Order.cpp]: Added and used.

File: universe/Universe.cpp (hunk 3 above)
Comment (LGM-Doyle) [reply universe/Universe.cpp:174]: Headers trimmed.

**Review by LGM-Doyle — COMMENTED** [review]
@adrianbroher the original code is old.  The least error prone method seemed to me to rebase and then apply the C++11 changes to the new code.

What would you like done?

**Comment by geoffthemedio** [comment]
````
5>       Link:
5>            Creating library ../../FreeOrion.lib and object ../../FreeOrion.exp
5>     1>
5>FleetWnd.obj : error LNK2001: unresolved external symbol "public: int __thiscall Pathfinder::NearestSystemTo(double,double)const " (?NearestSystemTo@Pathfinder@@QBEHNN@Z)
5>     1>
5>Common.lib(UniverseObject.obj) : error LNK2001: unresolved external symbol "public: int __thiscall Pathfinder::NearestSystemTo(double,double)const " (?NearestSystemTo@Pathfinder@@QBEHNN@Z)
5>     1>
5>Common.lib(ValueRef.obj) : error LNK2001: unresolved external symbol "public: int __thiscall Pathfinder::NearestSystemTo(double,double)const " (?NearestSystemTo@Pathfinder@@QBEHNN@Z)
5>     1>
5>MapWnd.obj : error LNK2001: unresolved external symbol "public: struct std::pair<class std::list<int,class std::allocator<int> >,double> __thiscall Pathfinder::ShortestPath(int,int,int)const " (?ShortestPath@Pathfinder@@QBE?AU?$pair@V?$list@HV?$allocator@H@std@@@std@@N@std@@HHH@Z)
5>     1>
5>Common.lib(Order.obj) : error LNK2001: unresolved external symbol "public: struct std::pair<class std::list<int,class std::allocator<int> >,double> __thiscall Pathfinder::ShortestPath(int,int,int)const " (?ShortestPath@Pathfinder@@QBE?AU?$pair@V?$list@HV?$allocator@H@std@@@std@@N@std@@HHH@Z)
5>     1>
5>Common.lib(Effect.obj) : error LNK2001: unresolved external symbol "public: struct std::pair<class std::list<int,class std::allocator<int> >,double> __thiscall Pathfinder::ShortestPath(int,int,int)const " (?ShortestPath@Pathfinder@@QBE?AU?$pair@V?$list@HV?$allocator@H@std@@@std@@N@std@@HHH@Z)
5>     1>
5>Common.lib(Fleet.obj) : error LNK2001: unresolved external symbol "public: struct std::pair<class std::list<int,class std::allocator<int> >,double> __thiscall Pathfinder::ShortestPath(int,int,int)const " (?ShortestPath@Pathfinder@@QBE?AU?$pair@V?$list@HV?$allocator@H@std@@@std@@N@std@@HHH@Z)
5>     1>
5>Common.lib(Universe.obj) : error LNK2001: unresolved external symbol "public: __thiscall Pathfinder::Pathfinder(void)" (??0Pathfinder@@QAE@XZ)
5>     1>
5>Common.lib(Universe.obj) : error LNK2001: unresolved external symbol "public: void __thiscall Pathfinder::InitializeSystemGraph(class std::vector<int,class std::allocator<int> >,int)" (?InitializeSystemGraph@Pathfinder@@QAEXV?$vector@HV?$allocator@H@std@@@std@@H@Z)
5>     1>
5>Common.lib(Universe.obj) : error LNK2001: unresolved external symbol "public: void __thiscall Pathfinder::UpdateEmpireVisibilityFilteredSystemGraphs(int)" (?UpdateEmpireVisibilityFilteredSystemGraphs@Pathfinder@@QAEXH@Z)
5>     1>
5>Common.lib(Condition.obj) : error LNK2001: unresolved external symbol "public: int __thiscall Pathfinder::JumpDistanceBetweenObjects(int,int)const " (?JumpDistanceBetweenObjects@Pathfinder@@QBEHHH@Z)
5>     1>
5>Common.lib(ValueRef.obj) : error LNK2001: unresolved external symbol "public: int __thiscall Pathfinder::JumpDistanceBetweenObjects(int,int)const " (?JumpDistanceBetweenObjects@Pathfinder@@QBEHHH@Z)
5>     1>
5>Common.lib(Fleet.obj) : error LNK2001: unresolved external symbol "public: bool __thiscall Pathfinder::SystemHasVisibleStarlanes(int,int)const " (?SystemHasVisibleStarlanes@Pathfinder@@QBE_NHH@Z)
5>     1>
5>Common.lib(ValueRef.obj) : error LNK2001: unresolved external symbol "public: double __thiscall Pathfinder::ShortestPathDistance(int,int)const " (?ShortestPathDistance@Pathfinder@@QBENHH@Z)
5>     1>
5>../../FreeOrion.exe : fatal error LNK1120: 8 unresolved externals
5>
5>Build FAILED.````

**Comment by LGM-Doyle** [comment]
I've push a new PR that manually adds Pathfinder to the msvc project files.  I thought that some portion of the new and improved CMake created the projects for msvc and OSX.

If this was the problem for msvc then the Xcode project may also need to be updated and how to do that was not immediately obvious from the Xcode projects.

If the problem is not with the msvc project files, then I don't understand from the error message that you provided how to solve this.  So, I have questions.

Windows I presume?

Does it compile/not compile `Pathfinder.cpp`?  Is it a compile/linker problem?

Did you try a clean build?

**Comment by geoffthemedio** [comment]
I didn't investigate, but if you didn't add Pathfinder.cpp to the MSVC project files, that's probably the problem.

The files need to go within a filter though, probably universe, not sitting at the top level.
solution

File: universe/Pathfinder.h [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+#include "UniverseObject.h"
+
+#include <boost/unordered_map.hpp>
+
+#include <vector>
+#include <map>
+#include <set>
+
+#ifdef FREEORION_WIN32
```
Comment (geoffthemedio) [inline universe/Pathfinder.h:12]: seems to build OK for me without this `#undef` block

Comment (geoffthemedio) [inline universe/Pathfinder.h]: seems to build OK for me without this `#undef` 
https://github.com/freeorion/freeorion/pull/1249/files#diff-1567159526cce829c7e2802017835646block

**Comment by geoffthemedio** [comment]
Builds / runs OK now.

**Comment by Vezzra** [comment]
Compiling `Pathfinder.cpp` fails. Build log for `Pathfinder.cpp`:
```
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:974:56: error: member reference type 'std::initializer_list<std::__1::shared_ptr<(anonymous namespace)::GraphImpl> >' is not a pointer; maybe you meant to use '.'?
        boost::get(vertex_system_id_t(), new_graph_impl->system_graph);
                                         ~~~~~~~~~~~~~~^~
                                                       .
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:974:58: error: no member named 'system_graph' in 'std::initializer_list<std::__1::shared_ptr<(anonymous namespace)::GraphImpl> >'
        boost::get(vertex_system_id_t(), new_graph_impl->system_graph);
                                         ~~~~~~~~~~~~~~  ^
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:977:54: error: member reference type 'std::initializer_list<std::__1::shared_ptr<(anonymous namespace)::GraphImpl> >' is not a pointer; maybe you meant to use '.'?
        boost::get(boost::edge_weight, new_graph_impl->system_graph);
                                       ~~~~~~~~~~~~~~^~
                                                     .
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:977:56: error: no member named 'system_graph' in 'std::initializer_list<std::__1::shared_ptr<(anonymous namespace)::GraphImpl> >'
        boost::get(boost::edge_weight, new_graph_impl->system_graph);
                                       ~~~~~~~~~~~~~~  ^
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:982:41: error: member reference type 'std::initializer_list<std::__1::shared_ptr<(anonymous namespace)::GraphImpl> >' is not a pointer; maybe you meant to use '.'?
        boost::add_vertex(new_graph_impl->system_graph);
                          ~~~~~~~~~~~~~~^~
                                        .
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:982:43: error: no member named 'system_graph' in 'std::initializer_list<std::__1::shared_ptr<(anonymous namespace)::GraphImpl> >'
        boost::add_vertex(new_graph_impl->system_graph);
                          ~~~~~~~~~~~~~~  ^
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:1012:85: error: member reference type 'std::initializer_list<std::__1::shared_ptr<(anonymous namespace)::GraphImpl> >' is not a pointer; maybe you meant to use '.'?
                boost::add_edge(system1_index, lane_dest_graph_index, new_graph_impl->system_graph);
                                                                      ~~~~~~~~~~~~~~^~
                                                                                    .
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:1012:87: error: no member named 'system_graph' in 'std::initializer_list<std::__1::shared_ptr<(anonymous namespace)::GraphImpl> >'
                boost::add_edge(system1_index, lane_dest_graph_index, new_graph_impl->system_graph);
                                                                      ~~~~~~~~~~~~~~  ^
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:1027:69: error: member reference type 'std::initializer_list<std::__1::shared_ptr<(anonymous namespace)::GraphImpl> >' is not a pointer; maybe you meant to use '.'?
    graph_changed = graph_changed || boost::num_edges(new_graph_impl->system_graph) != boost::num_edges(m_graph_impl->system_graph);
                                                      ~~~~~~~~~~~~~~^~
                                                                    .
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:1027:71: error: no member named 'system_graph' in 'std::initializer_list<std::__1::shared_ptr<(anonymous namespace)::GraphImpl> >'
    graph_changed = graph_changed || boost::num_edges(new_graph_impl->system_graph) != boost::num_edges(m_graph_impl->system_graph);
                                                      ~~~~~~~~~~~~~~  ^
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:1030:24: error: no member named 'swap' in 'std::initializer_list<std::__1::shared_ptr<(anonymous namespace)::GraphImpl> >'
        new_graph_impl.swap(m_graph_impl);
        ~~~~~~~~~~~~~~ ^
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:1060:64: error: no viable overloaded '='
            m_graph_impl->empire_system_graph_views[empire_id] = filtered_graph_ptr;
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~
In file included from /Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:1:
In file included from /Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.h:4:
In file included from /Users/user/SoftwareProjekte/FO/src/universe/UniverseObject.h:11:
In file included from /Users/user/SoftwareProjekte/FO/src/Xcode/dep/include/boost/signals2/signal.hpp:15:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:628:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3841:17: note: candidate function not viable: no known conversion from 'std::initializer_list<std::__1::shared_ptr<boost::filtered_graph<boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS, boost::property<SystemPathing::vertex_system_id_t, int, boost::property<boost::vertex_index_t, int, boost::no_property> >, boost::property<boost::edge_weight_t, double, boost::no_property>, boost::no_property, boost::listS>, (anonymous namespace)::GraphImpl::EdgeVisibilityFilter, boost::keep_all> > >' to 'const std::__1::shared_ptr<boost::filtered_graph<boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS, boost::property<SystemPathing::vertex_system_id_t, int, boost::property<boost::vertex_index_t, int, boost::no_property> >, boost::property<boost::edge_weight_t, double, boost::no_property>, boost::no_property, boost::listS>, (anonymous namespace)::GraphImpl::EdgeVisibilityFilter, boost::keep_all> >' for 1st argument
    shared_ptr& operator=(const shared_ptr& __r) _NOEXCEPT;
                ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3850:17: note: candidate function not viable: no known conversion from 'std::initializer_list<std::__1::shared_ptr<boost::filtered_graph<boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS, boost::property<SystemPathing::vertex_system_id_t, int, boost::property<boost::vertex_index_t, int, boost::no_property> >, boost::property<boost::edge_weight_t, double, boost::no_property>, boost::no_property, boost::listS>, (anonymous namespace)::GraphImpl::EdgeVisibilityFilter, boost::keep_all> > >' to 'std::__1::shared_ptr<boost::filtered_graph<boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS, boost::property<SystemPathing::vertex_system_id_t, int, boost::property<boost::vertex_index_t, int, boost::no_property> >, boost::property<boost::edge_weight_t, double, boost::no_property>, boost::no_property, boost::listS>, (anonymous namespace)::GraphImpl::EdgeVisibilityFilter, boost::keep_all> >' for 1st argument
    shared_ptr& operator=(shared_ptr&& __r) _NOEXCEPT;
                ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3848:9: note: candidate template ignored: could not match 'shared_ptr' against 'initializer_list'
        operator=(const shared_ptr<_Yp>& __r) _NOEXCEPT;
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3857:9: note: candidate template ignored: could not match 'shared_ptr' against 'initializer_list'
        operator=(shared_ptr<_Yp>&& __r);
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3865:9: note: candidate template ignored: could not match 'auto_ptr' against 'initializer_list'
        operator=(auto_ptr<_Yp>&& __r);
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3884:9: note: candidate template ignored: could not match 'unique_ptr' against 'initializer_list'
        operator=(unique_ptr<_Yp, _Dp>&& __r);
        ^
/Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:1070:64: error: no viable overloaded '='
            m_graph_impl->empire_system_graph_views[empire_id] = filtered_graph_ptr;
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~
In file included from /Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.cpp:1:
In file included from /Users/user/SoftwareProjekte/FO/src/universe/Pathfinder.h:4:
In file included from /Users/user/SoftwareProjekte/FO/src/universe/UniverseObject.h:11:
In file included from /Users/user/SoftwareProjekte/FO/src/Xcode/dep/include/boost/signals2/signal.hpp:15:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:628:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3841:17: note: candidate function not viable: no known conversion from 'std::initializer_list<std::__1::shared_ptr<boost::filtered_graph<boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS, boost::property<SystemPathing::vertex_system_id_t, int, boost::property<boost::vertex_index_t, int, boost::no_property> >, boost::property<boost::edge_weight_t, double, boost::no_property>, boost::no_property, boost::listS>, (anonymous namespace)::GraphImpl::EdgeVisibilityFilter, boost::keep_all> > >' to 'const std::__1::shared_ptr<boost::filtered_graph<boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS, boost::property<SystemPathing::vertex_system_id_t, int, boost::property<boost::vertex_index_t, int, boost::no_property> >, boost::property<boost::edge_weight_t, double, boost::no_property>, boost::no_property, boost::listS>, (anonymous namespace)::GraphImpl::EdgeVisibilityFilter, boost::keep_all> >' for 1st argument
    shared_ptr& operator=(const shared_ptr& __r) _NOEXCEPT;
                ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3850:17: note: candidate function not viable: no known conversion from 'std::initializer_list<std::__1::shared_ptr<boost::filtered_graph<boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS, boost::property<SystemPathing::vertex_system_id_t, int, boost::property<boost::vertex_index_t, int, boost::no_property> >, boost::property<boost::edge_weight_t, double, boost::no_property>, boost::no_property, boost::listS>, (anonymous namespace)::GraphImpl::EdgeVisibilityFilter, boost::keep_all> > >' to 'std::__1::shared_ptr<boost::filtered_graph<boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS, boost::property<SystemPathing::vertex_system_id_t, int, boost::property<boost::vertex_index_t, int, boost::no_property> >, boost::property<boost::edge_weight_t, double, boost::no_property>, boost::no_property, boost::listS>, (anonymous namespace)::GraphImpl::EdgeVisibilityFilter, boost::keep_all> >' for 1st argument
    shared_ptr& operator=(shared_ptr&& __r) _NOEXCEPT;
                ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3848:9: note: candidate template ignored: could not match 'shared_ptr' against 'initializer_list'
        operator=(const shared_ptr<_Yp>& __r) _NOEXCEPT;
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3857:9: note: candidate template ignored: could not match 'shared_ptr' against 'initializer_list'
        operator=(shared_ptr<_Yp>&& __r);
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3865:9: note: candidate template ignored: could not match 'auto_ptr' against 'initializer_list'
        operator=(auto_ptr<_Yp>&& __r);
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3884:9: note: candidate template ignored: could not match 'unique_ptr' against 'initializer_list'
        operator=(unique_ptr<_Yp, _Dp>&& __r);
        ^
```

**Comment by LGM-Doyle** [comment]
@Vezzra, thanks for the log file.  Could you try compiling this version?

I think that the problem is that Clang does not correctly compile list based initialization, which @adrianbroher pointed out.  I have removed it.

**Comment by Vezzra** [comment]
Build and runs fine on OSX now. I've also pushed the required changes to the Xcode project.

As far as I'm concerned, LGTM.

**Comment by geoffthemedio** [comment]
I don't know how to get a pull request onto this branch to work, but this commit moves the Pathfinder files into the universe filter in the MSVC project files: https://github.com/freeorion/freeorion/commit/489b9043aa6d43f8f202589e7d2d890350ffc2c2

**Comment by LGM-Doyle** [comment]
@geoffthemedio done.

I don't know how you are supposed to do it, but I just pulled my branch that @Vezzra had added a commit to and pulled and cherry picked your commit and then pushed.

@Vezzra, might have a simpler method, since he did add a commit to this branch.

**Comment by geoffthemedio** [comment]
I assume he just pushed directly to the branch, rather than making a pull request.

**Comment by Vezzra** [comment]
> I assume he just pushed directly to the branch, rather than making a pull request.

This.

It's a new github feature that has been around for some time, that allows you to push to PR branches even if they belong to the personal clones of the contributors (unless specifically disabled when the PR is created). Very handy! 😃

**Pull request merged**